provisioner/node_pools.go

```diff
@@ -123,13 +223,150 @@ func (p *AWSNodePoolProvisioner) Provision(ctx context.Context, nodePools []*api
+	if err != nil {
+		return "", err
+	}
+
+	args = append([]string{
+		"kubectl",
+		fmt.Sprintf("--server=%s", p.cluster.APIServerURL),
+		fmt.Sprintf("--token=%s", token.AccessToken),
+	}, args...)
+	if stdin != "" {
+		args = append(args, "-f", "-")
+	}
+
+	newApplyCommand := func() *exec.Cmd {
+		cmd := exec.Command(args[0], args[1:]...)
+		// prevent kubectl to find the in-cluster config
+		cmd.Env = []string{}
+		return cmd
+	}
+	var output string
+	applyManifest := func() error {
+		cmd := newApplyCommand()
+		if stdin != "" {
+			cmd.Stdin = strings.NewReader(stdin)
 		}
+		output, err = p.execManager.Run(ctx, p.logger, cmd)
+		return err
 	}
-	return true
+	err = backoff.Retry(applyManifest, backoff.WithMaxRetries(backoff.NewExponentialBackOff(), maxApplyRetries))
+	if err != nil {
+		return "", err
+	}
+	return output, nil
+}
+
+func (p *KarpenterNodePoolProvisioner) provisionNodePool(ctx context.Context, nodePool *api.NodePool, values map[string]interface{}) error {
+	template, err := p.generateNodePoolTemplate(nodePool, values, provisionersFileName)
+	if err != nil {
+		return err
+	}
+	if template == "" {
+		return nil
+	}
+	_, err = p.kubectlExecute(ctx, []string{"apply"}, template)
+	if err != nil {
+		return errors.Wrapf(err, "kubectl apply failed for node pool %s", nodePool.Name)
+	}
+	return nil
+}
+
+func (p *KarpenterNodePoolProvisioner) isKarpenterInstalled(ctx context.Context) (bool, error) {
+	gvr, err := kubernetes.ResolveKind(p.kubeMapper, crd)
+	if err != nil {
+		return false, err
+	}
+
+	_, err = p.kubeClient.Resource(gvr).Get(ctx, karpenterProvisionerResource, metav1.GetOptions{})
+	if err != nil {
+		serr, ok := err.(*apiErrors.StatusError)
+		if ok && serr.Status().Code == http.StatusNotFound {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+func (p *KarpenterNodePoolProvisioner) Reconcile(ctx context.Context, updater updatestrategy.UpdateStrategy) error {
+	karpenterPools := p.cluster.KarpenterPools()
+	installed, err := p.isKarpenterInstalled(ctx)
+	if err != nil {
+		return err
+	}
+	if !installed {
+		// skip if karpenter is not installed
+		p.logger.Infof("skipping reconcilation of karpenter node pools as karpenter is not installed")
+		return nil
+	}
+
+	provisionerGvr, err := kubernetes.ResolveKind(p.kubeMapper, karpenterProvisionerResource)
+	if err != nil {
+		return err
+	}
+	nodeTemplateGvr, err := kubernetes.ResolveKind(p.kubeMapper, karpenterAWSNodeTemplateResource)
+	if err != nil {
+		return err
+	}
+	existingProvisioners, err := p.kubeClient.Resource(provisionerGvr).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, pr := range existingProvisioners.Items {
+		if !inNodePoolList(&api.NodePool{Name: pr.GetName()}, karpenterPools) {
+			_ = p.kubeClient.Resource(provisionerGvr).Delete(ctx, pr.GetName(), metav1.DeleteOptions{})
+			_ = p.kubeClient.Resource(nodeTemplateGvr).Delete(ctx, pr.GetName(), metav1.DeleteOptions{})
```

Comment:
Ignoring the error doesn't feel good either. Let's use the old approach or at least print the error.

I was just worried about the case where the `provisioner` gets deleted but the `awstemplate` fails. It will never retry in this case and it will never be cleaned up. Not a big deal I guess.